src/librustc/hir/lowering.rs

```diff
@@ -862,9 +865,14 @@ impl<'a> LoweringContext<'a> {
                 PatKind::Wild => hir::PatKind::Wild,
                 PatKind::Ident(ref binding_mode, pth1, ref sub) => {
                     self.with_parent_def(p.id, |this| {
+                        let name = match this.resolver.get_resolution(p.id).map(|d| d.full_def()) {
+                            // Only pattern bindings are renamed
```

Comment:
I'm not sure I understand the question.
>At this stage we are post-parsing and post-name resolution, so where do the remaining ambiguities get resolved?

All ambiguities are resolved during name resolution, there's nothing ambiguous at this stage.

>Put another way, in some sense this renaming must happen somewhere, where does it happen (and how)?

Renaming is done twice, once during name resolution, its results aren't saved anywhere, and once more during lowering to HIR, its results are persistent and kept in HIR.
`Name`s in HIR need to be properly renamed to correctly treat cases like [this](https://github.com/rust-lang/rust/pull/29748#issuecomment-155532904), which in principle can be interpreted as requiring some rudimentary form of name resolution during type checking.